### ISSUE TITLE
DuckDB compilation errors

### ISSUE THREAD

**Issue by PedroTadim** [duckdb/duckdb#6964]
### What happens?

I can't get DuckDB to compile on machine using the latest sources:

With clang 15:
```
third_party/tpce-tool/tpce.cpp:24:2: error: reference to 'unique_ptr' is ambiguous
        unique_ptr<CBaseLoaderFactory> pLoaderFactory; // class factory that creates table loaders
        ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:242:11: note: candidate found by name lookup is 'std::unique_ptr'
    class unique_ptr
          ^
src/include/duckdb/common/unique_ptr.hpp:24:7: note: candidate found by name lookup is 'duckdb::unique_ptr'
class unique_ptr : public std::unique_ptr<_Tp, _Dp> {
      ^
third_party/tpce-tool/tpce.cpp:26:2: error: reference to 'unique_ptr' is ambiguous
        unique_ptr<CGenerateAndLoad> pGenerateAndLoad;
        ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:242:11: note: candidate found by name lookup is 'std::unique_ptr'
    class unique_ptr
          ^
src/include/duckdb/common/unique_ptr.hpp:24:7: note: candidate found by name lookup is 'duckdb::unique_ptr'
class unique_ptr : public std::unique_ptr<_Tp, _Dp> {
      ^
third_party/tpce-tool/tpce.cpp:49:21: error: reference to 'unique_ptr' is ambiguous
        pGenerateAndLoad = unique_ptr<CGenerateAndLoad>(
                           ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:242:11: note: candidate found by name lookup is 'std::unique_ptr'
    class unique_ptr
          ^
src/include/duckdb/common/unique_ptr.hpp:24:7: note: candidate found by name lookup is 'duckdb::unique_ptr'
class unique_ptr : public std::unique_ptr<_Tp, _Dp> {
      ^
In file included from third_party/tpce-tool/tpce.cpp:2:
In file included from src/include/duckdb/main/connection.hpp:12:
In file included from src/include/duckdb/common/serializer/buffered_file_writer.hpp:11:
In file included from src/include/duckdb/common/serializer.hpp:11:
In file included from src/include/duckdb/catalog/catalog.hpp:11:
In file included from src/include/duckdb/catalog/catalog_entry.hpp:11:
In file included from src/include/duckdb/common/common.hpp:12:
src/include/duckdb/common/helper.hpp:114:2: error: static assertion failed due to requirement 'sizeof(TPCE::DuckDBLoaderFactory) == 0': Use make_uniq instead of make_unique!
        static_assert(sizeof(T) == 0, "Use make_uniq instead of make_unique!");
        ^             ~~~~~~~~~~~~~~
third_party/tpce-tool/tpce.cpp:39:19: note: in instantiation of function template specialization 'duckdb::make_unique<TPCE::DuckDBLoaderFactory, duckdb::Connection &, std::basic_string<char> &, std::basic_string<char> &>' requested here
        pLoaderFactory = make_unique<DuckDBLoaderFactory>(con, schema, suffix);
                         ^
4 errors generated.
```
With gcc 11:
```
extension/sqlsmith/sqlsmith-extension.cpp: In function ‘duckdb::unique_ptr<duckdb::FunctionData> duckdb::ReduceSQLBind(duckdb::ClientContext&, duckdb::TableFunctionBindInput&, std::vector<duckdb::LogicalType>&, std::vector<std::__cxx11::basic_string<char> >&)’:
extension/sqlsmith/sqlsmith-extension.cpp:100:16: error: could not convert ‘result’ from ‘unique_ptr<duckdb::ReduceSQLFunctionData,default_delete<duckdb::ReduceSQLFunctionData>>’ to ‘unique_ptr<duckdb::FunctionData,default_delete<duckdb::FunctionData>>’
  100 |         return result;
      |                ^~~~~~
      |                |
      |                unique_ptr<duckdb::ReduceSQLFunctionData,default_delete<duckdb::ReduceSQLFunctionData>>
```
I also get this warning worth fixing:
```
extension/httpfs/s3fs.cpp: In static member function ‘static std::string duckdb::S3FileSystem::UrlDecode(std::string)’:
extension/httpfs/s3fs.cpp:106:66: warning: format ‘%x’ expects argument of type ‘unsigned int*’, but argument 3 has type ‘int*’ [-Wformat=]
  106 |                         sscanf(input.substr(i + 1, 2).c_str(), "%x", &ii);
```

### To Reproduce

Compile DuckDB

### OS:

Linux

### DuckDB Version:

Latest from sources

### DuckDB Client:

Shell

### Full Name:

Pedro Ferreira

### Affiliation:

Huawei

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by Mytherin** [comment]
Thanks for the report! I noticed the same issues and have fixed them as part of #6961

**Comment by Tishj** [comment]
I'm curious why our CI did not pick this up, or even when I compile all extensions + out of tree extensions together I don't run into this 🤔

**Comment by Mytherin** [comment]
`BUILD_TPCE` and `BUILD_SQLSMITH` are not enabled in the CI on the relevant boxes most likely

**Comment by PedroTadim** [comment]
I think I enabled the SQLsmith extension some time ago.

**Comment by Tishj** [comment]
Yea actually I think it might be because of the unity builds that I don't run into this

**Comment by PedroTadim** [comment]
Now this error appears:

benchmark/interpreted_benchmark.cpp: In member function ‘virtual duckdb::unique_ptr<duckdb::BenchmarkState> duckdb::InterpretedBenchmark::Initialize(duckdb::BenchmarkConfiguration&)’:
benchmark/interpreted_benchmark.cpp:385:16: error: could not convert ‘state’ from ‘unique_ptr<duckdb::InterpretedBenchmarkState>’ to ‘unique_ptr<duckdb::BenchmarkState>’
  385 |         return state;
      |                ^~~~~
      |                |
      |                unique_ptr<duckdb::InterpretedBenchmarkState>

**Comment by PedroTadim** [comment]
Now, this error appears:

benchmark/interpreted_benchmark.cpp: In member function ‘virtual duckdb::unique_ptr<duckdb::BenchmarkState> duckdb::InterpretedBenchmark::Initialize(duckdb::BenchmarkConfiguration&)’:
benchmark/interpreted_benchmark.cpp:385:16: error: could not convert ‘state’ from ‘unique_ptr<duckdb::InterpretedBenchmarkState>’ to ‘unique_ptr<duckdb::BenchmarkState>’
  385 |         return state;
      |                ^~~~~
      |                |
      |                unique_ptr<duckdb::InterpretedBenchmarkState>

**Comment by Tishj** [comment]
Hmm it seems we need to port some extra things from `std::unique_ptr`'s implementation to support implicit casting to base class

**Comment by Tishj** [comment]
> > Hmm it seems we need to port some extra things from `std::unique_ptr`'s implementation to support implicit casting to base class But I can't reproduce this locally `BUILD BENCHMARK=1 make reldebug` what are you compiling with?
> 
> gcc 11.3.0

That helps, but I was more wondering what your cmake settings looked like, that's why I provided that I can't reproduce it when I build with `BUILD_BENCHMARK=1 make reldebug`

**Comment by PedroTadim** [comment]
cmake -DCMAKE_INSTALL_PREFIX=$DuckDB_build_dir -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++ -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_FLAGS="-DDUCKDB_USE_STANDARD_ASSERT -DDUCKDB_CRASH_ON_ASSERT -fsanitize=address,undefined" -DCMAKE_CXX_FLAGS="-DDUCKDB_USE_STANDARD_ASSERT -DDUCKDB_CRASH_ON_ASSERT -fsanitize=address,undefined" -DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=address,undefined" -DBUILD_ICU_EXTENSION=ON -DBUILD_HTTPFS_EXTENSION=ON -DBUILD_PARQUET_EXTENSION=ON -DBUILD_TPCH_EXTENSION=ON -DBUILD_TPCDS_EXTENSION=ON -DBUILD_BENCHMARKS=ON -DBUILD_SQLSMITH=ON -DBUILD_TPCE=ON -DBUILD_JSON_EXTENSION=ON -DBUILD_SHELL=ON ../src

**Comment by Tishj** [comment]
You should really start using Ninja, because this compilation took forever, unless you have other settings that I'm missing (?)
But yea, even with
```
mkdir -p ./build/debug && \
	cd build/debug && \
	cmake -DCMAKE_INSTALL_PREFIX=$DuckDB_build_dir -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++ -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_FLAGS="-DDUCKDB_USE_STANDARD_ASSERT -DDUCKDB_CRASH_ON_ASSERT -fsanitize=address,undefined" -DCMAKE_CXX_FLAGS="-DDUCKDB_USE_STANDARD_ASSERT -DDUCKDB_CRASH_ON_ASSERT -fsanitize=address,undefined" -DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=address,undefined" -DBUILD_ICU_EXTENSION=ON -DBUILD_HTTPFS_EXTENSION=ON -DBUILD_PARQUET_EXTENSION=ON -DBUILD_TPCH_EXTENSION=ON -DBUILD_TPCDS_EXTENSION=ON -DBUILD_BENCHMARKS=ON -DBUILD_SQLSMITH=ON -DBUILD_TPCE=ON -DBUILD_JSON_EXTENSION=ON -DBUILD_SHELL=ON ../.. && \
	cmake --build . --config Debug
```
I can't reproduce this locally, so it's likely Linux related?

**Comment by Mytherin** [comment]
Should be fixed by #6967

**Comment by PedroTadim** [comment]
> You should really start using Ninja, because this compilation took forever, unless you have other settings that I'm missing (?) But yea, even with
> 
> ```
> mkdir -p ./build/debug && \
> 	cd build/debug && \
> 	cmake -DCMAKE_INSTALL_PREFIX=$DuckDB_build_dir -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++ -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_FLAGS="-DDUCKDB_USE_STANDARD_ASSERT -DDUCKDB_CRASH_ON_ASSERT -fsanitize=address,undefined" -DCMAKE_CXX_FLAGS="-DDUCKDB_USE_STANDARD_ASSERT -DDUCKDB_CRASH_ON_ASSERT -fsanitize=address,undefined" -DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=address,undefined" -DBUILD_ICU_EXTENSION=ON -DBUILD_HTTPFS_EXTENSION=ON -DBUILD_PARQUET_EXTENSION=ON -DBUILD_TPCH_EXTENSION=ON -DBUILD_TPCDS_EXTENSION=ON -DBUILD_BENCHMARKS=ON -DBUILD_SQLSMITH=ON -DBUILD_TPCE=ON -DBUILD_JSON_EXTENSION=ON -DBUILD_SHELL=ON ../.. && \
> 	cmake --build . --config Debug
> ```
> 
> I can't reproduce this locally, so it's likely Linux related?
> 
> Or maybe because I'm using a newer version of gcc
> 
> ```
> ➜  duckdb git:(master) ✗ gcc --version 
> Apple clang version 14.0.0 (clang-1400.0.29.202)
> Target: arm64-apple-darwin22.3.0
> Thread model: posix
> InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
> ```

You are using clang? Yes, I am going to start using ninja now. I just never noticed a noticeable improvement over make a few years back. Meanwhile, I can no longer reproduce this, so I can close the issue.
